### PR TITLE
feat: support `date_trunc` function

### LINKED ISSUE #6301: feat: support `date_trunc` function
### Is your feature request related to a problem? Please describe.

Mz: https://materialize.com/docs/sql/functions/date-trunc/
Pg: https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC:~:text=9.9.2.-,date_trunc,-The%20function%20date_trunc

We should support it cuz this is useful in temporal data (And also required by some BI Tools).

Just ask, cc @TennyZhuang , do we already support part of this or similar?

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR adds support for `date_trunc` function.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* SQL commands, functions, and operators

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

As documented in postgres: https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC.

Except that `timezone` must be specified if `source` is a timestamp with time zone:
```
date_trunc(field, TIMESTAMPZ, time_zone) // ok
date_trunc(field, TIMESTAMPZ) // not supported for now
```
Reason: https://github.com/risingwavelabs/risingwave/pull/6365#discussion_r1022400476.

## Refer to a related PR or issue link (optional)
resolve #6301

### CODE REVIEW

File: src/frontend/src/expr/type_inference/func.rs [hunk 1]
```diff
@@ -816,6 +816,9 @@ fn build_type_derive_map() -> FuncSigMap {
     map.insert(E::ToTimestamp, vec![T::Float64], T::Timestampz);
     map.insert(E::AtTimeZone, vec![T::Timestamp, T::Varchar], T::Timestampz);
     map.insert(E::AtTimeZone, vec![T::Timestampz, T::Varchar], T::Timestamp);
+    map.insert(E::DateTrunc, vec![T::Varchar, T::Timestamp], T::Timestamp);
+    map.insert(E::DateTrunc, vec![T::Varchar, T::Timestampz], T::Timestampz);
```
Comment (xiangjinwu) [inline src/frontend/src/expr/type_inference/func.rs:820]: Actually, we should only allow `[T::Varchar, T::Timestampz, T::Varchar]` and forbid its binary variant:
> When the input value is of type timestamp with time zone, the truncation is performed with respect to a particular time zone; for example, truncation to day produces a value that is midnight in that zone. By default, truncation is done with respect to the current [TimeZone](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-TIMEZONE) setting, but the optional time_zone argument can be provided to specify a different time zone.

As of now we do not support implicit TimeZone setting and require explicit zone parameter in all such operations. #5826

**Review by BowenXiao1999 — APPROVED** [review]
LSTM!

File: src/expr/src/expr/expr_binary_nonnull.rs [hunk 2]
```diff
@@ -383,6 +386,40 @@ fn build_at_time_zone_expr(
...
+        DataType::Timestamp => Box::new(BinaryExpression::<
+            Utf8Array,
+            NaiveDateTimeArray,
+            NaiveDateTimeArray,
+            _,
+        >::new(l, r, ret, date_trunc_timestamp)),
+        DataType::Timestampz => Box::new(BinaryExpression::<
+            Utf8Array,
+            NaiveDateTimeArray,
+            NaiveDateTimeArray,
```
Comment (xiangjinwu) [inline src/expr/src/expr/expr_binary_nonnull.rs:404]: `Timestampz` is represented as an `i64`, and its interpretation of year/../minute depends on timezone. So the tenary truncation operation would be a chained operation of the following 3:
* timestampz AT TIME ZONE zone -> timestamp
* truncate(field, timestamp) -> timestamp
* timestamp AT TIME ZONE zone -> timestampz

File: src/expr/src/vector_op/date_trunc.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    ts: NaiveDateTimeWrapper,
+) -> Result<NaiveDateTimeWrapper> {
+    date_trunc_timestamp(field, ts)
+}
+
+#[inline]
+pub fn date_trunc_interval(field: &str, interval: IntervalUnit) -> Result<IntervalUnit> {
+    Ok(match field {
+        "microseconds" => interval,
+        "milliseconds" => interval,
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/date_trunc.rs:51]: Let's add a placeholder for `truncate_millis` here. #4514

**Review by xiangjinwu — COMMENTED** [review]
rest lgtm

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**